src/utils/wallet/bitget_wallet.ts

```
@@ -0,0 +1,174 @@
+  constructor() {
+    super();
+  }
+
+  connectWallet = async (): Promise<any> => {
+    // check whether there is an Bitget Wallet extension
+    if (!this.provider) {
+      throw new Error("Bitget Wallet extension not found");
+    }
+
+    try {
+      await this.provider?.requestAccounts() // Connect to Bitget Wallet extension
+      await this.provider?.switchNetwork('signet')
+    } catch (error) {
+      if ((error as Error)?.message?.includes('rejected')) {
+        throw new Error('Connection to Bitget Wallet was rejected')
+      } else {
+        throw new Error((error as Error)?.message)
+      }
+    }
+
+    const address = await this.getAddress()
+    const publicKeyHex = await this.getPublicKeyHex()
+
+    if (address && publicKeyHex) {
+      this.bitgetWalletInfo = {
+        publicKeyHex,
+        address,
+      };
+      return this;
+    } else {
+      throw new Error("Could not connect to Bitget Wallet");
+    }
+  };
+
+  getWalletProviderName = async (): Promise<string> => {
+    return 'Bitget Wallet';
+  };
+
+  getAddress = async (): Promise<string> => {
+    let accounts = (await this.provider?.getAccounts()) || []
+    return accounts[0]
+  };
+
+  getPublicKeyHex = async (): Promise<string> => {
+    let publicKey = await this.provider?.getPublicKey()
+    if (!publicKey) {
+      console.log('bitget Wallet not connected')
+      return ''
+    }
+    return publicKey
+  };
+
+  signPsbt = async (psbtHex: string): Promise<string> => {
+    const data = {
+      method: 'signPsbt',
+      params: {
+        from: this.provider?.selectedAddress,
+        __internalFunc: '__signPsbt_babylon',
+        psbtHex,
+        options: {
+          autoFinalized: true
+        }
+      }
+    }
+
+    let signedPsbt = await this.provider?.request('dappsSign', data)
+    let psbt = Psbt.fromHex(signedPsbt)
+    try {
+      psbt.finalizeAllInputs()
+    } catch (error) {
+      console.log('babylon-signPsbt failed', error)
```

This should throw an error

---

The key point is to ensure that all PSBTs are finalized, and we do not want it to throw an error if the PSBT has already been finalized. Therefore, if this procedure is not needed, the try-catch block could be removed, but there is no guarantee that the PSBT will be finalized.

---

If `finalizeAllInputs()` throws an error won't this mean that the psbt will eitherway not be finalized? If we care about the case where the psbt has already been finalized can we check this some other way instead of catching an exception here?

---

We can achieve this in other ways, but there are scenarios in which the finalize process might fail in practice. For example, in a 2-of-3 multisig transaction, a PSBT might fail to finalize if only one signature has been collected. In such cases, the partially signed PSBT can be sent to the next signer for further processing even if finallized failed. We use try-catch to accommodate these scenarios.

Does the babylon Dapp handle such cases? Or can we ensure that the PSBT here must be successfully finalized here, and if not, an exception should be thrown?

---

Hmm I see. Let's expect the transaction to be successfully finalized here but not your concern in a comment. If the transaction is not finalized here then the rest of the application will break as it expects that the transaction is fully signed after this step.

---

Got it. I will change the logic here to ensure every PSBT is fully finalized, and if not, an error will be thrown.

---

Done.